### PR TITLE
Add token balance subgraph

### PR DESCRIPTION
**Pull request by jorgeavaldez**
- Light refactoring of the existing code
- Added the `FreeNFTBalance` entity so we can track ownership of NFTs created by the factory contract
- Trigger Transfer indexing from the new edition handler

For now we're not displaying the actual transfers from the frontend. If we want to do so in the future, we can always add them here.

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -10,4 +10,18 @@ type FreeNFTDrop @entity(immutable: true) {
...
+  balances: [FreeNFTBalance!]! @derivedFrom(field: "drop")
+}
+
+
+type FreeNFTBalance @entity(immutable: false) {
+  "The owner is <drop address>-<owner address>"
+  id: String! # address
+
+  address: Bytes! # owner address
+  quantity: BigInt! # address
```
Comment (0xkarmacoma) [inline schema.graphql]: `# address` <- stale comment?

File: schema.graphql [hunk 2]
```diff
@@ -10,4 +10,18 @@ type FreeNFTDrop @entity(immutable: true) {
   description: String
   animationUrl: String
   imageUrl: String
+  
+  balances: [FreeNFTBalance!]! @derivedFrom(field: "drop")
+}
+
+
+type FreeNFTBalance @entity(immutable: false) {
+  "The owner is <drop address>-<owner address>"
```
Comment (0xkarmacoma) [inline schema.graphql]: I don't understand this, do you mean this instead?

`"The id is <drop address>-<owner address>"`

File: schema.graphql (hunk 1 above)
Comment (jorgeavaldez) [reply schema.graphql]: yes, changing!

File: schema.graphql (hunk 2 above)
Comment (jorgeavaldez) [reply schema.graphql]: yes!

File: src/handlers/handle-transfers.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    return;
+  }
+
+  let editionAddress = event.address;
+  
+  let drop = FreeNFTDrop.load(Bytes.fromHexString(editionAddress.toHexString()));
+  
+  if (drop == null) {
+    log.critical(
```
Comment (0xkarmacoma) [inline src/handlers/handle-transfers.ts]: wouldn't that happen for every transfer from a non-Showtime drop? `log.critical` terminates the subgraph, I think we should either just ignore it or `log.debug` and move on

File: src/handlers/handle-transfers.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+let NULL_ADDRESS = "0x0000000000000000000000000000000000000000";
+
+
+function updateBalance(
+  event: TransferEvent, 
+  drop: FreeNFTDrop, 
+  address: Address, 
+  isSender: boolean = false,
+): void {
```
Comment (0xkarmacoma) [inline src/handlers/handle-transfers.ts]: consider this a mega-nit if you will but
- I'm not a big fan of the `isSender` bool to change behavior
- most of `updateBalance` is concerned with actually loading/creating a balance, I think it makes sense to break it out

I would do it like this:

```ts
function loadBalance(
  drop: FreeNFTDrop,
  address: Address,
): FreeNFTBalance | null {
  if (address.toHexString() == NULL_ADDRESS) {
    // we don't track balances for the null address
    return null;
  }

  let id = drop.id.toHexString() + "-" + address.toHexString();
  let balance = FreeNFTBalance.load(id);

  if (balance == null) {
    balance = new FreeNFTBalance(id);

    balance.address = address;
    balance.quantity = BigInt.zero();
    balance.drop = drop.id;
  }

  if (balance.address != address) {
    log.error(
      "Balance with id {} has owner address {}, which does not match address {}",
      [
        balance.id,
        balance.address.toHexString(),
        address.toHexString(),
      ]
    );

    return null;
  }

  return balance;
}
```

```ts
function updateBalance(
  balance: FreeNFTBalance | null,
  quantityDelta: i32,
  event: TransferEvent,
): void {
  if (balance == null) {
    return;
  }

  balance.quantity = balance.quantity.plus(BigInt.fromI32(quantityDelta));

  if (balance.quantity.lt(BigInt.zero())) {
    log.error(
      "Balance with id {} has negative quantity {} after transfer in tx {}",
      [
        balance.id,
        balance.quantity.toString(),
        event.transaction.hash.toHexString(),
      ],
    );

    return;
  }

  balance.updatedAt = event.block.timestamp;
  balance.save();
}
```

In `handleTransfer`:

```
  updateBalance(loadBalance(drop, sender), -1, event);
  updateBalance(loadBalance(drop, receiver), 1, event);
```

File: src/handlers/handle-transfers.ts (hunk 3 above)
Comment (jorgeavaldez) [reply src/handlers/handle-transfers.ts]: no! the handler is only triggered when a drop is created and if it matches the parent drop. not sure how that works but it's some template magic from the graph!

File: src/handlers/handle-transfers.ts (hunk 4 above)
Comment (jorgeavaldez) [reply src/handlers/handle-transfers.ts]: 100% agree, this is much, much nicer! making the change!

File: src/handlers/handle-transfers.ts (hunk 3 above)
Comment (jorgeavaldez) [reply src/handlers/handle-transfers.ts]: for posterity here's the docs on templates https://thegraph.com/docs/en/developing/creating-a-subgraph/#data-source-templates-for-dynamically-created-contracts

File: subgraph.yaml [hunk 5]
```diff
@@ -28,4 +28,25 @@ dataSources:
...
+      kind: ethereum/events
+      apiVersion: 0.0.7
+      language: wasm/assemblyscript
+      entities:
+        - FreeNFTBalance
+      abis:
+        - name: SingleEditionMintable
+          file: ./abis/SingleEditionMintable.json
+      eventHandlers:
+        - event: Transfer(indexed address,indexed address,indexed uint256)
```
Comment (0xkarmacoma) [inline subgraph.yaml:49]: TIL the `uint256 tokenId` is indexed (I just double checked) 👌

File: src/handlers/handle-transfers.ts (hunk 3 above)
Comment (0xkarmacoma) [reply src/handlers/handle-transfers.ts]: very cool!

docs: https://thegraph.com/docs/en/developing/creating-a-subgraph/#data-source-templates-for-dynamically-created-contracts

**Review by 0xkarmacoma — APPROVED** [review]

**Comment by 0xkarmacoma** [comment]
Thanks for the clarifications! I'll let you merge

**Pull request merged**